common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java

```diff
@@ -84,57 +103,162 @@ private CoverageSearchEvent getLastEvent(CoveragePeriod period) {
     }
 
     @Override
-    public CoverageSearchEvent insertCoverage(int searchId, long searchEventId, Collection<String> beneficiaryIds) {
+    public CoverageSearchEvent insertCoverage(int periodId, long searchEventId, List<String> beneficiaryIds) {
 
-        CoveragePeriod period = findCoveragePeriod(searchId);
+        // Make sure that coverage period and searchEvent actually exist in the database before inserting
+        findCoveragePeriod(periodId);
         CoverageSearchEvent searchEvent = findCoverageSearchEvent(searchEventId);
 
-        Collection<Coverage> coverages = beneficiaryIds.stream()
-                .map(bene -> buildCoverage(period, searchEvent, bene)).collect(toList());
+        int written = 0;
+        while (written < beneficiaryIds.size()) {
 
-        JobStatus status = getSearchStatus(searchId);
-        if (status == JobStatus.IN_PROGRESS) {
-            coverageRepo.saveAll(coverages);
-        } else {
-            throw new InvalidJobAccessException("cannot update coverages if coverage search is not in progress (current status " + status + " )");
+            try (Connection connection = dataSource.getConnection();
+                 PreparedStatement statement = connection.prepareStatement(INSERT)) {
+
+                int stopCondition = written + BATCH_INSERT_SIZE > beneficiaryIds.size() ? beneficiaryIds.size() : written + BATCH_INSERT_SIZE;
+
+                for (int index = written; index < stopCondition; index++) {
```

Comment:
Having a loop surrounding a try with a loop within yields varying **levels of abstraction**.  My preference would be to refactor the contents of the try block to a private method that would be descriptively named aiding in readability.